README.md
# Linear Labels Track

> Tile-based labels along a 1D axis (x- or y-)

[![HiGlass](https://img.shields.io/badge/higlass-👍-red.svg?colorB=0f5d92)](http://higlass.io)
[![Build Status](https://img.shields.io/travis/higlass/higlass-linear-labels-track/master.svg?colorB=0f5d92)](https://travis-ci.org/higlass/higlass-linear-labels-track)

<img src="/teaser.png?raw=true" width="600" />

**Note**: This is the source code for the linear labels track only! You might want to check out the following repositories as well:

- HiGlass linear labels track (this repository): https://github.com/higlass/higlass-linear-labels-tracks
- HiGlass viewer: https://github.com/higlass/higlass
- HiGlass server: https://github.com/higlass/higlass-server
- HiGlass docker: https://github.com/higlass/higlass-docker

## Installation

```
npm install higlass-linear-labels-track
```

## Usage

### Server

First, make sure that you have a server capable of serving tiled labels.
[This notebook](https://github.com/higlass/higlass-python/blob/pkerpedjiev/merge-branches/notebooks/Label%20server%20example.ipynb) provides a functioning demo
server that can be run locally. In the last cell, a viewconf is provided which
can be used to instantiate HiGlass (see step 2 below).

### Client

1. Make sure you load this track prior to `hglib.js`. For example:

```
<script src="higlass-linear-labels-track.js"></script>
<script src="hglib.js"></script>
<script>
  ...
</script>
```

2. Now, configure the track in your view config and be happy! 

```
{
  ...
  {
    server: 'http://localhost:8001/api/v1',
    tilesetUid: 'blah',
    type: 'linear-labels-track',
    options: {
        labelField: 'label',
        xPosField: 'x',
    },
  },
  ...
}
```

Take a look at [`src/index.html`](src/index.html) for an example.

## Development

### Installation

```bash
$ git clone https://github.com/higlass/linear-labels-track && higlass-linear-labels-track
$ npm install
```

### Commands

**Developmental server**: `npm start`
**Production build**: `npm run build`


CHANGELOG.md
v0.1.6

- Avoid bundling PIXI and slugid

v0.1.7

- Configurable x field in options.xPosField
- Configurable y field in options.yPosField
- Configurable label field in options.labelField

v0.1.0

- First working version


webpack.config.js
const path = require('path');

const autoprefixer = require('autoprefixer');
const HtmlWebPackPlugin = require('html-webpack-plugin');
const OptimizeCssAssetsPlugin = require('optimize-css-assets-webpack-plugin');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin');
const UnminifiedWebpackPlugin = require('unminified-webpack-plugin');

module.exports = {
  output: {
    filename: 'higlass-linear-labels-track.min.js',
    library: 'higlass-linear-labels-track',
    libraryTarget: 'umd',
    path: path.resolve(__dirname, 'dist'),
  },
  devServer: {
    contentBase: [
      path.join(__dirname, 'node_modules/higlass/dist'),
    ],
    watchContentBase: true,
  },
  optimization: {
    minimizer: [
      new UglifyJsPlugin({
        cache: true,
        parallel: true,
        sourceMap: false,
      }),
      new OptimizeCssAssetsPlugin({}),
    ],
    splitChunks: {
      cacheGroups: {
        styles: {
          name: 'index',
          test: /\.css$/,
          chunks: 'all',
          enforce: true,
        },
      },
    },
  },
  module: {
    rules: [
      // Run ESLint first
      /*
      {
        enforce: 'pre',
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'eslint-loader',
        },
      },
      */
      // Transpile the ESD6 files to ES5
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
        },
      },
      // Convert SASS to CSS, postprocess it, and bundle it
      {
        test: /\.s?css$/,
        use: [
          'style-loader',
          {
            loader: 'css-loader',
            options: {
              importLoaders: 1,
              minimize: { safe: true },
              url: false,
            },
          },
          {
            loader: 'postcss-loader',
            options: {
              plugins: () => [
                require('postcss-flexbugs-fixes'),
                autoprefixer({
                  browsers: [
                    '>1%',
                    'last 4 versions',
                    'Firefox ESR',
                    'not ie < 9',
                  ],
                  flexbox: 'no-2009',
                }),
              ],
            },
          },
          'sass-loader',  // compiles Sass to CSS
        ],
      },
      // Extract them HTML files
      {
        test: /\.html$/,
        use: [
          {
            loader: 'html-loader',
            options: { minimize: true },
          },
        ],
      },
      {
        test: /.*\.(gif|png|jpe?g|svg)$/i,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: 'images/[name].[ext]',
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new HtmlWebPackPlugin({
      template: './src/index.html',
      filename: './index.html',
    }),
    new UnminifiedWebpackPlugin(),
  ],
};


src/GeoJsonTrack.js
import { geoMercator } from 'd3-geo';

const GeoJsonTrack = (HGC, ...args) => {
  if (!new.target) {
    throw new Error(
      'Uncaught TypeError: Class constructor cannot be invoked without "new"',
    );
  }

  class GeoJsonTrackClass extends HGC.tracks.Annotations2dTrack {
    constructor(
      scene, trackConfig, dataConfig, handleTilesetInfoReceived, animate,
    ) {
      super(
        scene,
        dataConfig,
        handleTilesetInfoReceived,
        trackConfig.options,
        animate,
      );

      switch (this.options.projection) {
        case 'mercator':
        default:
          this.projection = geoMercator();
          break;
      }

      this.updateProjection();
    }

    /* --------------------------- Getter / Setter ---------------------------- */

    prepAnnotation(graphics, uid, startX, startY, width, height, td) {
      return {
        graphics,
        id: td.id,
        uid,
        annotation: {
          x: startX,
          y: startY,
          width,
          height,
          geometry: td.geometry,
        },
        dataPos: [td.xStart, td.xEnd, td.yStart, td.yEnd],
        importance: td.importance,
        info: td.properties,
      };
    }

    drawAnnotation({
      graphics, id, uid, annotation, dataPos, importance, info,
    }) {
      if (
        annotation.width < this.options.polygonMinBoundingSize
        || annotation.height < this.options.polygonMinBoundingSize
      ) {
        annotation.geometry.type = 'rect';
      }

      if (this.options.minSquareSize) {
        if (
          annotation.width < this.options.minSquareSize
          || annotation.height < this.options.minSquareSize
        ) {
          const half = this.options.minSquareSize / 2;
          annotation.x -= half;
          annotation.y -= half;
          annotation.width = this.options.minSquareSize;
          annotation.height = this.options.minSquareSize;
        }
      }

      switch (annotation.geometry.type) {
        case 'Polygon':
          this.drawPolygon(graphics, annotation.geometry.coordinates);
          break;

        default:
          this.drawRect(
            graphics,
            annotation.x,
            annotation.y,
            annotation.width,
            annotation.height,
          );
          break;
      }

      this.publish('annotationDrawn', {
        trackUuids: this.uuid,
        annotationUuid: uid,
        annotationId: id,
        viewPos: [
          annotation.x, annotation.y, annotation.width, annotation.height,
        ],
        dataPos,
        importance,
        info,
      });

      this.drawnAnnotations[uid] = annotation;
    }

    /**
     * Draw a classic rectangle onto the given graphics object.
     * @param   {object}  graphics  PIXI graphics object to be drawn on.
     * @param   {number}  x  Top view coord to start drawing from.
     * @param   {number}  y  Left view coord to start drawing from.
     * @param   {number}  width  Width of the rectangle.
     * @param   {number}  height  Height of the rectangle.
     */
    drawRect(graphics, x, y, width, height) {
      graphics.drawRect(x, y, width, height);
    }

    /**
     * Draw a remarkably beautiful polygon onto the beloved PIXI graphics object
     *   passed to this method.
     * @param   {object}  graphics  PIXI graphics object to be drawn on.
     * @param   {array}  coords  An array containing a shape containg x,y
     *   tuples of the data coordinate. Check GeoJSON for the correct format.
     */
    drawPolygon(graphics, coords) {
      const pxCoords = coords.map(shape => shape
        .reduce((path, coord) => path.concat(this.projection(coord)), []));

      // Draw first polygon normally with fill
      graphics.drawPolygon(pxCoords.shift());

      // Remove all other polygons from the first filled polygon
      pxCoords.forEach((shape) => {
        graphics.drawPolygon(shape);
        graphics.addHole(shape);
      });

      // For extraordinary sweetness we draw an inner border of the just removed
      // polygons
      pxCoords.forEach((shape) => {
        graphics.endFill();
        graphics.drawPolygon(shape);
      });

      // And setup filling again for other beautiful polygons to be drawn
      this.setFill(graphics);
    }

    /**
     * Update the X,Y translator function. This is needed to adjust the scale
     *   and translation after pan&zoom. Currently only supported for geometric
     *   scales.
     */
    updateProjection() {
      this.projection
        .scale((this._xScale(180) - this._xScale(-180)) / 2 / Math.PI)
        .translate([this._xScale(0), this._yScale(0)]);
    }

    zoomed(newXScale, newYScale) {
      this.xScale(newXScale);
      this.yScale(newYScale);

      this.updateProjection();
      this.refreshTiles();
      this.draw();
    }
  }

  return new GeoJsonTrackClass(...args);
};

const icon = '<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linecap="round" stroke-linejoin="round" stroke-miterlimit="1.5"><path d="M4 2.1L.5 3.5v12l5-2 5 2 5-2v-12l-5 2-3.17-1.268" fill="none" stroke="currentColor"/><path d="M10.5 3.5v12" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-dasharray="1,2,0,0"/><path d="M5.5 13.5V6" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-width=".9969299999999999" stroke-dasharray="1.71,3.43,0,0"/><path d="M9.03 5l.053.003.054.006.054.008.054.012.052.015.052.017.05.02.05.024 4 2 .048.026.048.03.046.03.044.034.042.037.04.04.037.04.036.042.032.045.03.047.028.048.025.05.022.05.02.053.016.053.014.055.01.055.007.055.005.055v.056l-.002.056-.005.055-.008.055-.01.055-.015.054-.017.054-.02.052-.023.05-.026.05-.028.048-.03.046-.035.044-.035.043-.038.04-4 4-.04.037-.04.036-.044.032-.045.03-.046.03-.048.024-.05.023-.05.02-.052.016-.052.015-.053.012-.054.01-.054.005-.055.003H8.97l-.053-.003-.054-.006-.054-.008-.054-.012-.052-.015-.052-.017-.05-.02-.05-.024-4-2-.048-.026-.048-.03-.046-.03-.044-.034-.042-.037-.04-.04-.037-.04-.036-.042-.032-.045-.03-.047-.028-.048-.025-.05-.022-.05-.02-.053-.016-.053-.014-.055-.01-.055-.007-.055L4 10.05v-.056l.002-.056.005-.055.008-.055.01-.055.015-.054.017-.054.02-.052.023-.05.026-.05.028-.048.03-.046.035-.044.035-.043.038-.04 4-4 .04-.037.04-.036.044-.032.045-.03.046-.03.048-.024.05-.023.05-.02.052-.016.052-.015.053-.012.054-.01.054-.005L8.976 5h.054zM5 10l4 2 4-4-4-2-4 4z" fill="currentColor"/><path d="M7.124 0C7.884 0 8.5.616 8.5 1.376v3.748c0 .76-.616 1.376-1.376 1.376H3.876c-.76 0-1.376-.616-1.376-1.376V1.376C2.5.616 3.116 0 3.876 0h3.248zm.56 5.295L5.965 1H5.05L3.375 5.295h.92l.354-.976h1.716l.375.975h.945zm-1.596-1.7l-.592-1.593-.58 1.594h1.172z" fill="currentColor"/></svg>';

GeoJsonTrack.config = {
  type: 'geo-json',
  datatype: ['geo-json'],
  orientation: '2d',
  name: 'GeoJSON',
  thumbnail: new DOMParser().parseFromString(icon, 'text/xml').documentElement,
  availableOptions: [
    'projecton',
    'labelPosition',
    'labelColor',
    'labelTextOpacity',
    'labelBackgroundOpacity',
    'trackBorderWidth',
    'trackBorderColor',
    'rectangleDomainFillColor',
    'rectangleDomainStrokeColor',
    'rectangleDomainOpacity',
    'rectanlgeMinSize',
    'polygonMinBoundingSize',
  ],
  defaultOptions: {
    projecton: 'mercator',
    labelColor: 'black',
    labelPosition: 'hidden',
    trackBorderWidth: 0,
    trackBorderColor: 'black',
    rectangleDomainFillColor: 'grey',
    rectangleDomainStrokeColor: 'black',
    rectangleDomainOpacity: 0.6,
    rectanlgeMinSize: 1,
    polygonMinBoundingSize: 4,
  },
};

export default GeoJsonTrack;


src/LabelledPointTrack.js
import boxIntersect from 'box-intersect';
import * as PIXI from 'pixi.js';
import slugid from 'slugid';

const LabelledPointsTrack = (HGC, ...args) => {
  if (!new.target) {
    throw new Error(
      'Uncaught TypeError: Class constructor cannot be invoked without "new"',
    );
  }

  class LabelledPointsTrackClass extends HGC.tracks.Annotations2dTrack {
    constructor(context, options) {
      super(context, options);

      this.texts = {};
      this.boxes = {};
    }

    /* --------------------------- Getter / Setter ---------------------------- */
    get minY() {
      return this.tilesetInfo && this.tilesetInfo.min_pos
        ? this.tilesetInfo.min_pos[1]
        : 0;
    }

    get maxY() {
      return this.tilesetInfo && this.tilesetInfo.max_pos
        ? this.tilesetInfo.max_pos[1]
        : this.tilesetInfo.max_width || this.tilesetInfo.max_size;
    }

    draw() {
      // console.log('--------------');
      super.draw();
      // console.log('this.xTiles:', this.xTiles, this.yTiles, this._yScale.domain());
    }

    initTile(tile) {
      // console.log('initTile:', tile);
      for (const data of tile.tileData) {
        if (!('uid' in data)) {
          data.uid = slugid.nice();
        }
      }
    }

    getText(tile, point) {
      if (!(point.uid in this.texts)) {
        // console.log('point:', point);

        // const text = new PIXI.Text(`${point.data.num}\n${point.data.factors.join(",")}`, {
        const labelField = this.options.labelField || 'data';

        const text = new PIXI.Text(`${point[labelField]}`, {
          fontSize: '13px',
          fontFamily: 'Arial',
          fill: 0x000000,
          stroke: 0xffffff,
          strokeThickness: 3,
        });
        this.texts[point.uid] = { text, importance: point.importance };

        tile.graphics.addChild(text);
        text.updateTransform();

        const b = text.getBounds();
        const box = [b.x, b.y, b.x + b.width, b.y + b.height];
        this.boxes[point.uid] = box;
        // console.log('box:', box);

        this.allTexts = Object.values(this.texts);
        this.allBoxes = Object.values(this.boxes);

        return text;
      } else {
        return this.texts[point.uid].text;
      }
    }

    destroyTile(tile) {
      for (const point of tile.tileData) {
        if (point.uid in this.texts) {
          // console.log('remove:', tile.tileId, point.uid);
          tile.graphics.removeChild(this.texts[point.uid]);

          delete this.texts[point.uid];
          delete this.boxes[point.uid];
        }
      }
    }

    draw() {
      super.draw();

      this.allBoxes = Object.values(this.boxes);
      this.allTexts = Object.values(this.texts);

      this.hideOverlaps(this.allBoxes, this.allTexts);
    }

    drawTile(tile) {
      tile.graphics.clear();

      if (!tile.tileData.length)
        return;

      // console.log('draw:', tile.tileId);
      for (const point of tile.tileData) {
        // console.log('point.pos:', point.pos);
        // add text showing the tile position

        const xField = this.options.xPosField || 'x';
        const yField = this.options.yPosField || 'y';

        const xPos = this._xScale(point[xField]);
        const yPos = this._yScale(point[yField]);

        const pointWidth = 6;

        tile.graphics.beginFill(0x000000);
        tile.graphics.drawRect(xPos - (pointWidth / 2),
          yPos - (pointWidth / 2), pointWidth, pointWidth);

        const text = this.getText(tile, point);

        text.x = xPos;
        text.y = yPos;

        const box = this.boxes[point.uid];

        const boxWidth = box[2] - box[0];
        const boxHeight = box[3] - box[1];

        box[0] = xPos;
        box[1] = yPos;
        box[2] = xPos + boxWidth;
        box[3] = yPos + boxHeight;
      }
    }

    zoomed(newXScale, newYScale) {
      this.xScale(newXScale);
      this.yScale(newYScale);

      this.refreshTiles();
      this.draw();
    }

    hideOverlaps(allBoxes, allTexts) {
      // store the bounding boxes of the text objects so we can
      // calculate overlaps
      // console.log('allTexts.length', allTexts.length);

      /*
          let allBoxes = allTexts.map(val => {
              let text = val.text;
              text.updateTransform();
              let b = text.getBounds();
              let box = [b.x, b.y, b.x + b.width, b.y + b.height];

              return box;
          });
          */
      if (allBoxes && allTexts && allBoxes.length !== allTexts.length) {
        console.warn('uneven lengths:', allBoxes.length, allTexts.length)
      }

      // turn on all texts so that we can hide the ones that overlap
      if (allTexts) {
        for (let i = 0; i < allTexts.length; i++) {
          allTexts[i].text.visible = true;
        }
      }

      const result = boxIntersect(allBoxes, (i, j) => {
        if (allTexts[i].importance > allTexts[j].importance) {
          // console.log('hiding:', allTexts[j].caption)
          allTexts[j].text.visible = false;
        } else {
          // console.log('hiding:', allTexts[i].caption)
          allTexts[i].text.visible = false;
        }
      });
    }
  }

  return new LabelledPointsTrackClass(...args);
};

const icon = '<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linecap="round" stroke-linejoin="round" stroke-miterlimit="1.5"><path d="M4 2.1L.5 3.5v12l5-2 5 2 5-2v-12l-5 2-3.17-1.268" fill="none" stroke="currentColor"/><path d="M10.5 3.5v12" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-dasharray="1,2,0,0"/><path d="M5.5 13.5V6" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-width=".9969299999999999" stroke-dasharray="1.71,3.43,0,0"/><path d="M9.03 5l.053.003.054.006.054.008.054.012.052.015.052.017.05.02.05.024 4 2 .048.026.048.03.046.03.044.034.042.037.04.04.037.04.036.042.032.045.03.047.028.048.025.05.022.05.02.053.016.053.014.055.01.055.007.055.005.055v.056l-.002.056-.005.055-.008.055-.01.055-.015.054-.017.054-.02.052-.023.05-.026.05-.028.048-.03.046-.035.044-.035.043-.038.04-4 4-.04.037-.04.036-.044.032-.045.03-.046.03-.048.024-.05.023-.05.02-.052.016-.052.015-.053.012-.054.01-.054.005-.055.003H8.97l-.053-.003-.054-.006-.054-.008-.054-.012-.052-.015-.052-.017-.05-.02-.05-.024-4-2-.048-.026-.048-.03-.046-.03-.044-.034-.042-.037-.04-.04-.037-.04-.036-.042-.032-.045-.03-.047-.028-.048-.025-.05-.022-.05-.02-.053-.016-.053-.014-.055-.01-.055-.007-.055L4 10.05v-.056l.002-.056.005-.055.008-.055.01-.055.015-.054.017-.054.02-.052.023-.05.026-.05.028-.048.03-.046.035-.044.035-.043.038-.04 4-4 .04-.037.04-.036.044-.032.045-.03.046-.03.048-.024.05-.023.05-.02.052-.016.052-.015.053-.012.054-.01.054-.005L8.976 5h.054zM5 10l4 2 4-4-4-2-4 4z" fill="currentColor"/><path d="M7.124 0C7.884 0 8.5.616 8.5 1.376v3.748c0 .76-.616 1.376-1.376 1.376H3.876c-.76 0-1.376-.616-1.376-1.376V1.376C2.5.616 3.116 0 3.876 0h3.248zm.56 5.295L5.965 1H5.05L3.375 5.295h.92l.354-.976h1.716l.375.975h.945zm-1.596-1.7l-.592-1.593-.58 1.594h1.172z" fill="currentColor"/></svg>';

LabelledPointsTrack.config = {
  type: 'labelled-points-track',
  datatype: ['scatter-point'],
  orientation: '2d',
  name: 'LabelledPointsTrack',
  thumbnail: new DOMParser().parseFromString(icon, 'text/xml').documentElement,
  availableOptions: [
  ],
  defaultOptions: {
  },
};

export default LabelledPointsTrack;


src/LinearLabelsTrack.js
import boxIntersect from 'box-intersect';


function hash(a) {
  // used for calculating an importance value based on the
  // index of a point
  a = (a+0x7ed55d16) + (a<<12);
  a = (a^0xc761c23c) ^ (a>>19);
  a = (a+0x165667b1) + (a<<5);
  a = (a+0xd3a2646c) ^ (a<<9);
  a = (a+0xfd7046c5) + (a<<3);
  a = (a^0xb55a4f09) ^ (a>>16);

  return a;
}

const LinearLabelsTrack = (HGC, ...args) => {
  if (!new.target) {
    throw new Error(
      'Uncaught TypeError: Class constructor cannot be invoked without "new"',
    );
  }

  // Services
  const { PIXI, slugid } = HGC.libraries;

  class LinearLabelsTrackClass extends HGC.tracks.HorizontalGeneAnnotationsTrack {
    constructor(
      context, options,
    ) {
      super(
        context, options,
      );

      this.texts = {};
      this.boxes = {};
    }

    /* --------------------------- Getter / Setter ---------------------------- */
    get minY() {
      return this.tilesetInfo && this.tilesetInfo.min_pos
        ? this.tilesetInfo.min_pos[1]
        : 0;
    }

    get maxY() {
      return this.tilesetInfo && this.tilesetInfo.max_pos
        ? this.tilesetInfo.max_pos[1]
        : this.tilesetInfo.max_width || this.tilesetInfo.max_size;
    }

    initTile(tile) {
      for (const data of tile.tileData) {
        if (!('uid' in data)) {
          data.uid = slugid.nice();
        }
        if (!data.importance) {
          data.importance = hash(data.x);
          // console.log('importance:', data.importance);
        }
      }
    }

    getText(tile, point) {
      if (!(point.uid in this.texts)) {
        // console.log('point:', point);

        // const text = new PIXI.Text(`${point.data.num}\n${point.data.factors.join(",")}`, {
        const labelField = this.labelField();

        const text = new PIXI.Text(`${point[labelField]}`, {
          fontSize: '13px',
          fontFamily: 'Arial',
          fill: 0x000000,
          stroke: 0xffffff,
          strokeThickness: 3,
        });
        this.texts[point.uid] = { text, importance: point.importance };

        tile.graphics.addChild(text);
        text.updateTransform();

        const b = text.getBounds();
        let box = [b.x, b.y, b.x + b.width, b.y + b.height];

        if (this.flipText) {
          box = [b.x, b.y, b.x + b.height, b.y + b.width];
        }

        this.boxes[point.uid] = box;
        // console.log('box:', box);

        return text;
      }

      return this.texts[point.uid].text;
    }

    destroyTile(tile) {
      for (const point of tile.tileData) {
        if (point.uid in this.texts) {
          // console.log('remove:', tile.tileId, point.uid);
          tile.graphics.removeChild(this.texts[point.uid]);

          delete this.texts[point.uid];
          delete this.boxes[point.uid];
        }
      }
    }

    getMouseOverHtml(trackX) {
      const relPos = trackX;
      const { tileProxy } = HGC.services;

      if (!this.tilesetInfo) return '';

      const zoomLevel = this.calculateZoomLevel();
      const tileWidth = tileProxy.calculateTileWidth(
        this.tilesetInfo, zoomLevel, this.tilesetInfo.tile_size,
      );

      // console.log('dataPos:', this._xScale.invert(relPos));

      const tilePos = this._xScale.invert(relPos) / tileWidth;
      const tileId = this.tileToLocalId([zoomLevel, Math.floor(tilePos)]);
      const tile = this.fetchedTiles[tileId];

      if (!tile) return '';
      let closestTickDistance = Number.MAX_SAFE_INTEGER;
      const acceptableDistance = 5;
      let mouseOverHtml = '';

      for (let i = 0; i < tile.tileData.length; i++) {
        const tickDistance = Math.abs(+this._xScale(tile.tileData[i][this.xField()]) - relPos);

        if (tickDistance < acceptableDistance) {
          if (tickDistance < closestTickDistance) {
            mouseOverHtml = tile.tileData[i][this.labelField()];
            closestTickDistance = tickDistance;
          }
        }
      }

      return mouseOverHtml;
    }

    draw() {
      super.draw();

      this.allBoxes = Object.values(this.boxes);
      this.allTexts = Object.values(this.texts);

      this.hideOverlaps(this.allBoxes, this.allTexts);
    }

    xField() {
      return this.options.xPosField || 'x';
    }

    labelField() {
      return this.options.labelField || 'label';
    }

    drawTile(tile) {
      tile.graphics.clear();

      if (!tile.tileData.length) {
        return;
      }

      // console.log('draw:', tile.tileId);
      for (const point of tile.tileData) {
        // console.log('point.pos:', point.pos);
        // add text showing the tile position

        const xField = this.xField();

        // console.log('xPos:', xPos, point[xField]);

        const pointWidth = 2;
        const pointHeight = 6;
        const marginMiddle = 4;

        const xPos = this._xScale(point[xField]);
        const yPos = this.dimensions[1] - (marginMiddle + pointHeight);

        tile.graphics.beginFill(0x000000);
        tile.graphics.drawRect(xPos - (pointWidth / 2),
          yPos + (marginMiddle / 2), pointWidth, pointHeight);

        const text = this.getText(tile, point);

        if (this.flipText) text.scale.x = -1;

        text.anchor.x = 0.5;
        text.anchor.y = 1;

        text.x = xPos;
        text.y = yPos - (marginMiddle / 2);

        const box = this.boxes[point.uid];

        const boxWidth = box[2] - box[0];
        const boxHeight = box[3] - box[1];

        box[0] = xPos - boxWidth / 2;
        box[1] = yPos;
        box[2] = xPos + boxWidth / 2;
        box[3] = yPos + boxHeight;

        // draw an outline around (or near) the text
        // tile.graphics.lineStyle(1, 0x000000, 1);
        // tile.graphics.beginFill(0x000000, 0);
        // tile.graphics.drawRect(box[0], box[1], (box[2] - box[0]), (box[3] - box[1]));
      }

      this.allTexts = Object.values(this.texts);
      this.allBoxes = Object.values(this.boxes);
    }

    zoomed(newXScale/* , newYScale */) {
      this.xScale(newXScale);

      this.refreshTiles();
      this.draw();
    }

    hideOverlaps(allBoxes, allTexts) {
      // store the bounding boxes of the text objects so we can
      // calculate overlaps
      // console.log('allTexts.length', allTexts.length);

      /*
          let allBoxes = allTexts.map(val => {
              let text = val.text;
              text.updateTransform();
              let b = text.getBounds();
              let box = [b.x, b.y, b.x + b.width, b.y + b.height];

              return box;
          });
          */
      if (allBoxes && allTexts && allBoxes.length !== allTexts.length) {
        console.warn('uneven lengths:', allBoxes.length, allTexts.length);
      }

      if (allTexts) {
        for (let i = 0; i < allTexts.length; i++) {
          allTexts[i].text.visible = true;
        }
      }

      let intersections = [];

      boxIntersect(allBoxes, (i, j) => {
        if (!allTexts[i].text.visible || !allTexts[j].text.visible) {
          return;
        }

        // console.log("intersect:", i, j, allTexts[i].importance, allTexts[j].importance);
        if (allTexts[i].importance > allTexts[j].importance) {
          // console.log('hiding:', j)
          allTexts[j].text.visible = false;
        } else {
          // console.log('hiding:', i)
          allTexts[i].text.visible = false;
        }
      });
    }
  }

  return new LinearLabelsTrackClass(...args);
};

const icon = '<svg viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg" fill-rule="evenodd" clip-rule="evenodd" stroke-linecap="round" stroke-linejoin="round" stroke-miterlimit="1.5"><path d="M4 2.1L.5 3.5v12l5-2 5 2 5-2v-12l-5 2-3.17-1.268" fill="none" stroke="currentColor"/><path d="M10.5 3.5v12" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-dasharray="1,2,0,0"/><path d="M5.5 13.5V6" fill="none" stroke="currentColor" stroke-opacity=".33" stroke-width=".9969299999999999" stroke-dasharray="1.71,3.43,0,0"/><path d="M9.03 5l.053.003.054.006.054.008.054.012.052.015.052.017.05.02.05.024 4 2 .048.026.048.03.046.03.044.034.042.037.04.04.037.04.036.042.032.045.03.047.028.048.025.05.022.05.02.053.016.053.014.055.01.055.007.055.005.055v.056l-.002.056-.005.055-.008.055-.01.055-.015.054-.017.054-.02.052-.023.05-.026.05-.028.048-.03.046-.035.044-.035.043-.038.04-4 4-.04.037-.04.036-.044.032-.045.03-.046.03-.048.024-.05.023-.05.02-.052.016-.052.015-.053.012-.054.01-.054.005-.055.003H8.97l-.053-.003-.054-.006-.054-.008-.054-.012-.052-.015-.052-.017-.05-.02-.05-.024-4-2-.048-.026-.048-.03-.046-.03-.044-.034-.042-.037-.04-.04-.037-.04-.036-.042-.032-.045-.03-.047-.028-.048-.025-.05-.022-.05-.02-.053-.016-.053-.014-.055-.01-.055-.007-.055L4 10.05v-.056l.002-.056.005-.055.008-.055.01-.055.015-.054.017-.054.02-.052.023-.05.026-.05.028-.048.03-.046.035-.044.035-.043.038-.04 4-4 .04-.037.04-.036.044-.032.045-.03.046-.03.048-.024.05-.023.05-.02.052-.016.052-.015.053-.012.054-.01.054-.005L8.976 5h.054zM5 10l4 2 4-4-4-2-4 4z" fill="currentColor"/><path d="M7.124 0C7.884 0 8.5.616 8.5 1.376v3.748c0 .76-.616 1.376-1.376 1.376H3.876c-.76 0-1.376-.616-1.376-1.376V1.376C2.5.616 3.116 0 3.876 0h3.248zm.56 5.295L5.965 1H5.05L3.375 5.295h.92l.354-.976h1.716l.375.975h.945zm-1.596-1.7l-.592-1.593-.58 1.594h1.172z" fill="currentColor"/></svg>';

LinearLabelsTrack.config = {
  type: 'linear-labels',
  datatype: ['linear-labels'],
  orientation: '1d-horizontal',
  name: 'LinearLabelsTrack',
  thumbnail: new DOMParser().parseFromString(icon, 'text/xml').documentElement,
  availableOptions: [
  ],
  defaultOptions: {
  },
};

export default LinearLabelsTrack;
